### ISSUE TITLE
Deploy does not pull latest submodule updates & overrites manual submodule updates with older version

### ISSUE THREAD

**Issue by securityvoid** [projectkudu/kudu#2314]
I've got a Azure Functions app that is deploying via continuous deployment. I've pushed updates to the submodule, but these are not pulling on subsequent deploys. In fact, its overwriting the newly updated submodule on subsequent deploys :-P.

If I navigate to the scm console and run:
```git submodule update --remote```

This successfully updates my submodule to the latest updates. However, every time I update the code on my main application and the application redeploys, it overwrites the submodule updates with code which is now 2-3 commits behind the latest submodule commit.

Any idea what is going on, and how to fix this?

The code it keeps putting back in place is the same submodule code that existed the first time the submodule was added. As a result it seems like its being "cached".

Is there a manual way to remove this cache?

**Comment by davidebbo** [comment]
Kudu runs the following logic to update the submodules:

```
git submodule sync
submodule update --init --recursive --force
```

Try running the same commands manually from Kudu console to check if it behaves any differently.

**Comment by securityvoid** [comment]
@davidebbo The combination of those two commands reproduces the issue and overwrites the existing file(s). 

I'm not a "git" expert, but I think you need the --remote in order for it to update the latest changes:

[From their manual:](https://git-scm.com/docs/git-submodule#git-submodule---remote)
```--remote
This option is only valid for the update command. Instead of using the superproject’s recorded SHA-1 to update the submodule, use the status of the submodule’s remote-tracking branch. The remote used is branch’s remote (branch.<name>.remote), defaulting to origin. The remote branch used defaults to master, but the branch name may be overridden by setting the submodule.<name>.branch option in either .gitmodules or .git/config (with .git/config taking precedence).

This works for any of the supported update procedures (--checkout, --rebase, etc.). The only change is the source of the target SHA-1. For example, submodule update --remote --merge will merge upstream submodule changes into the submodules, while submodule update --merge will merge superproject gitlink changes into the submodules.

In order to ensure a current tracking branch state, update --remote fetches the submodule’s remote repository before calculating the SHA-1. If you don’t want to fetch, you should use submodule update --remote --no-fetch.

Use this option to integrate changes from the upstream subproject with your submodule’s current HEAD. Alternatively, you can run git pull from the submodule, which is equivalent except for the remote branch name: update --remote uses the default upstream repository and submodule.<name>.branch, while git pull uses the submodule’s branch.<name>.merge. Prefer submodule.<name>.branch if you want to distribute the default upstream branch with the superproject and branch.<name>.merge if you want a more native feel while working in the submodule itself.```

**Comment by securityvoid** [comment]
Normally I'd just add something like this to my deploy script. However, in this case, the sub-project is my deploy script. :-(

**Comment by davidebbo** [comment]
That sounds like a different scenario. No git expert either, but my understanding is that the superproject always keeps track of the sha1 that it is currently using for the submodule. If your submodule changes, you need to take an explicit step in your superproject to start getting the changes (by making a commit is superproject). The idea is that if you checkout any commit in the superproject, you always get the right submodule commit that matches that superproject commit.

By contrast, if I understand correctly, `--remote` seems to say that you want the very latest commit from the submodule, regardless of what the superproject says. Is this really what you want? It's definitely not something we'd want to do by default, as it would break a lot of scenarios.

**Comment by securityvoid** [comment]
@davidebbo I actually did want the latest version to be pulled automatically, but the more I think about it that is definitely silly. I definitely understand how that would cause problems as a general rule.

I found the following guide which has a section on [updating submodules](https://chrisjean.com/git-submodules-adding-using-removing-and-updating/), which makes it clear what needs to be done.

Essentially its:
1. git submodule init
2. git submodule update
3. cd SUBMODULE_FOLDER
4. git checkout master
5. git pull
6. cd ../
7. git add .
8. git commit

I've done this, and I'm doing a new deploy now. I expect this will solve my issue.

Sorry for bothering you. This is clearly due to my ignorance on how submodules are supposed to work, rather than any issue with Azure.
